kroxylicious-additional-filters/kroxylicious-encryption/src/main/java/io/kroxylicious/filter/encryption/EnvelopeEncryptionFilter.java

```diff
@@ -0,0 +1,228 @@
+                throw new IllegalStateException();
+            }
+            result.add(maybeDecodeRecords(partitionData, (MemoryRecords) partitionData.records()));
+        }
+        return join(result);
+    }
+
+    private CompletableFuture<PartitionData> maybeDecodeRecords(PartitionData fpr, MemoryRecords memoryRecords) {
+        final CompletableFuture<PartitionData> result;
+        if (!anyRecordSmellsOfEncryption(memoryRecords)) {
+            result = CompletableFuture.completedFuture(fpr);
+        }
+        else {
+            ByteBuffer buffer = null; // TODO
+            MemoryRecordsBuilder builder = recordsBuilder(buffer, memoryRecords);
+            List<CompletableFuture<Integer>> futures = new ArrayList<>();
+            for (var kafkaRecord : memoryRecords.records()) {
+                if (!smellsOfEncryption(kafkaRecord.value())) {
+                    builder.append(kafkaRecord.timestamp(), kafkaRecord.key(), kafkaRecord.value(), kafkaRecord.headers());
+                    futures.add(CompletableFuture.completedFuture(1));
+                }
+                else {
+                    ByteBuffer value = kafkaRecord.value();
+                    var cf = dekCache.resolve(value).thenApply((DekContext<K> dekContext) -> {
+
+                        ByteBuffer output = null;
+                        try {
+                            int plaintextSize = kafkaRecord.valueSize();
+                            // TODO ^^ this is an overestimate!
+                            output = bufferPool.acquire(plaintextSize);
+                            dekContext.decode(value, output);
+                            builder.append(kafkaRecord.timestamp(), kafkaRecord.key(), output, kafkaRecord.headers());
+                            return 1;
+                        }
+                        finally {
+                            bufferPool.release(output);
+                        }
+                    }).toCompletableFuture();
+                    futures.add(cf);
+                }
+            }
+            result = join(futures)
+                    .thenApply(ignored -> builder.build())
+                    .thenApply(fpr::setRecords);
+        }
+        return result;
+    }
+
+    private boolean anyRecordSmellsOfEncryption(MemoryRecords memoryRecords) {
+        return StreamSupport.stream(memoryRecords.records().spliterator(), false)
+                .anyMatch(kafkaRecord -> smellsOfEncryption(kafkaRecord.value()));
+    }
+
+    private boolean smellsOfEncryption(ByteBuffer value) {
+        return true; // TODO implement this
+    }
+
+    private static MemoryRecordsBuilder recordsBuilder(ByteBuffer buffer, MemoryRecords records) {
+        RecordBatch firstBatch = records.firstBatch();
+        return new MemoryRecordsBuilder(buffer,
+                firstBatch.magic(),
+                firstBatch.compressionType(), // TODO we might not want to use the client's compression
```

Comment:
ah.. you even had a TODO about it :)

Reply:
Question: Should we handle the compression using a pair of filters the bracket the encryption filter, or should the encryption filter handle it all itself? I guess the latter might have better performance if we're only processing buffers once in a more streaming way.

Reply:
I think I make it a responsibility of the EnvelopeEncryptionFilter.